packages/core/lib/copy.js

```diff
@@ -12,51 +12,53 @@ var cpr_options = {
 // won't override individual files. If a file exists, it will
 // simply move onto the next file.
 
-var copy = function(from, to, extra_options, callback) {
-  if (typeof extra_options === "function") {
-    callback = extra_options;
-    extra_options = {};
-  }
+const copy = function(from, to, extra_options) {
+  const options = _.merge(_.clone(cpr_options), extra_options);
 
-  var options = _.merge(_.clone(cpr_options), extra_options);
-
-  cpr(from, to, options, function(err, files) {
-    var new_files = [];
+  return new Promise((resolve, reject) => {
+    cpr(from, to, options, function(err, files) {
+      if (err) {
+        reject(err);
+      }
+      var new_files = [];
 
-    // Remove placeholders. Placeholders allow us to copy "empty" directories,
-    // but lets NPM and git not ignore them.
-    files = files || [];
-    for (var file of files) {
-      if (file.match(/.*PLACEHOLDER.*/) != null) {
-        fs.unlinkSync(file);
-        continue;
+      // Remove placeholders. Placeholders allow us to copy "empty" directories,
+      // but lets NPM and git not ignore them.
+      files = files || [];
+      for (let file of files) {
+        if (file.match(/.*PLACEHOLDER.*/) != null) {
+          fs.unlinkSync(file);
+          continue;
+        }
+        new_files.push(file);
       }
-      new_files.push(file);
-    }
 
-    callback(err, new_files);
+      resolve(new_files);
+    });
   });
 };
 
-copy.file = function(from, to, callback) {
-  var readStream = fs.createReadStream(from, "utf8");
-  var writeStream = fs.createWriteStream(to, "utf8");
+copy.file = function(from, to) {
+  return new Promise((resolve, reject) => {
+    const readStream = fs.createReadStream(from, "utf8");
+    const writeStream = fs.createWriteStream(to, "utf8");
 
-  readStream.on("error", function(err) {
-    callback(err);
-    callback = function() {};
-  });
+    readStream.on("error", function(err) {
+      reject(err);
+      reject = function() {};
+    });
 
-  writeStream.on("error", function(err) {
-    callback(err);
-    callback = function() {};
-  });
+    writeStream.on("error", function(err) {
+      reject(err);
+      reject = function() {};
+    });
 
-  writeStream.on("finish", function() {
-    callback();
-  });
+    writeStream.on("finish", function() {
+      resolve();
+    });
 
-  readStream.pipe(writeStream);
+    readStream.pipe(writeStream);
+  });
```

Comment:
Instead of overwriting `reject` you should use a pattern like (untested):

```
function handleFinish() {
  resolve();
}
function handleError(err) {
   readStream.removeListener("error", handleError);
   writeStream.removeListener("error", handleError);
   writeStream.removeListener("finish", handleFinish);
   reject(err);
}
readStream.once("error", handleError);
readStream.once("error", handleError);
writeStream.once("finish", handleFinish);

readStream.pipe(writeStream);
```

This will ensure `resolve` and `reject` can never be called multiple times.

Reply:
Yeah this seems like the right kind of patter for this.